File: 03_Backend/app/models/audit_log.py
from datetime import datetime, timezone

from sqlalchemy import DateTime, String, Text, text
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import BaseModel


class AuditLog(BaseModel):
    __tablename__ = "audit_logs"

    timestamp: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        index=True,
    )
    user_uuid: Mapped[str | None] = mapped_column(String(36), index=True, default=None)
    user_name: Mapped[str] = mapped_column(String(200))
    user_role: Mapped[str] = mapped_column(String(50))
    action: Mapped[str] = mapped_column(String(100), index=True)
    entity_type: Mapped[str | None] = mapped_column(String(100), default=None)
    entity_uuid: Mapped[str | None] = mapped_column(String(36), default=None)
    entity_name: Mapped[str | None] = mapped_column(String(300), default=None)
    description: Mapped[str | None] = mapped_column(Text, default=None)
    http_method: Mapped[str | None] = mapped_column(String(10), default=None)
    endpoint: Mapped[str | None] = mapped_column(String(500), default=None)
    ip_address: Mapped[str | None] = mapped_column(String(45), default=None)
    user_agent: Mapped[str | None] = mapped_column(String(500), default=None)
    status: Mapped[str] = mapped_column(String(20), default="SUCCESS")
    metadata_json: Mapped[str | None] = mapped_column(Text, default=None)

    def __repr__(self) -> str:
        return f"<AuditLog {self.uuid} [{self.action}] by {self.user_name}>"

File: mobile_app/lib/features/admin/audit/domain/entities/audit_log.dart
class AuditLog {
  final int id;
  final String uuid;
  final DateTime timestamp;
  final String? userUuid;
  final String userName;
  final String userRole;
  final String action;
  final String? entityType;
  final String? entityUuid;
  final String? entityName;
  final String? description;
  final String? httpMethod;
  final String? endpoint;
  final String? ipAddress;
  final String? userAgent;
  final String status;
  final String? metadataJson;
  final DateTime createdAt;
  final DateTime updatedAt;

  const AuditLog({
    required this.id,
    required this.uuid,
    required this.timestamp,
    this.userUuid,
    required this.userName,
    required this.userRole,
    required this.action,
    this.entityType,
    this.entityUuid,
    this.entityName,
    this.description,
    this.httpMethod,
    this.endpoint,
    this.ipAddress,
    this.userAgent,
    required this.status,
    this.metadataJson,
    required this.createdAt,
    required this.updatedAt,
  });
}

class AuditLogFilterValues {
  final List<String> actions;
  final List<String> entityTypes;

  const AuditLogFilterValues({
    required this.actions,
    required this.entityTypes,
  });
}

class AuditLogExportItem {
  final DateTime timestamp;
  final String userName;
  final String userRole;
  final String action;
  final String? entityType;
  final String? entityName;
  final String? description;
  final String status;

  const AuditLogExportItem({
    required this.timestamp,
    required this.userName,
    required this.userRole,
    required this.action,
    this.entityType,
    this.entityName,
    this.description,
    required this.status,
  });
}
